mlrun/api/db/sqldb/db.py

```diff
@@ -3446,3 +3491,117 @@ def _is_background_task_timeout_exceeded(background_task_record) -> bool:
+        run_uid: str,
+        project: str,
+        iter: int = 0,
+    ):
+        run = self._get_run(session, run_uid, project, iter)
+        if not run:
+            raise mlrun.errors.MLRunNotFoundError(
+                f"Run not found: uid={run_uid}, project={project}, iter={iter}"
+            )
+
+        run_notifications = {
+            notification.name: notification
+            for notification in self._find_notifications(session, run_id=run.id)
+        }
+        notifications = []
+        for notification_model in notification_models:
+            new_notification = False
+            notification = run_notifications.get(notification_model.name, None)
+            if not notification:
+                new_notification = True
+                notification = Notification(
+                    name=notification_model.name, run=run.id, project=project
+                )
+
+            notification.kind = notification_model.kind
+            notification.message = notification_model.message
+            notification.severity = notification_model.severity
+            notification.when = ",".join(notification_model.when)
+            notification.condition = notification_model.condition
+            notification.status = (
+                notification_model.status
+                or mlrun.utils.notifications.NotificationStatus.PENDING.value
+            )
+            notification.sent_time = notification_model.sent_time
+
+            k8s = mlrun.api.utils.singletons.k8s.get_k8s()
+            if k8s.running_inside_kubernetes_cluster and not (
+                notification_model.params and "secret" in notification_model.params
+            ):
+                # store params as secret
+                secret_name = f"notification-{run_uid}-{notification.name}"
+                secret_data = notification_model.params
+                k8s.store_secrets(secret_name, secret_data)
+                notification.params = {"secret": secret_name}
+            else:
+                notification.params = notification_model.params
+
+            logger.debug(
+                f"Storing {'new' if new_notification else 'existing'} notification",
+                notification_name=notification.name,
+                run_uid=run_uid,
+                project=project,
+            )
+            notifications.append(notification)
+
+        self._upsert(session, notifications)
+
+    def list_notifications(
+        self,
+        session,
+        run_uid: str,
+        project: str = "",
+        iter: int = 0,
+    ) -> typing.List[mlrun.model.Notification]:
+        run = self._get_run(session, run_uid, project, iter)
+        if not run:
+            return []
+
+        return [
+            self._transform_notification_record_to_schema(notification)
+            for notification in self._query(session, Notification, run=run.id).all()
+        ]
+
+    def delete_notifications(
```

Comment:
This is currently only called when deleting a project. Should at least be called when a run is deleted (the DB won't automatically delete dependencies).

Reply:
By using in the DB Run model - `notifications = relationship(Notification, cascade="all, delete-orphan")`
This makes it so the DB *will* automatically delete the notifications

Reply:
Spoke with @theSaarco offline and verified - with mysql this works, in sqlite the feature requires a config which isn't set by default (https://stackoverflow.com/questions/5890250/on-delete-cascade-in-sqlite3). So leaving as is.